### PR TITLE
NAS-102283 / 12 / Run periodic tasks when system is ready 

### PR DESCRIPTION
**Pull request by sonicaj**
This commit adds the ability to execute periodic tasks when the system is ready. Motivation for this is tasks which require network interaction, as middlewared boots early in the process, networking isn't setup. We still support  attribute but if  is enabled, that takes precedence over.

### CODE REVIEW

**Comment by sonicaj** [comment]
I did consider that but then didn't go through because there might be tasks which don't require the system to be ready like processing alerts ( i think it doesn't require network ? ). If we do that, then these tasks would have to wait until the system has completely booted.

**Comment by william-gr** [comment]
I dont think thats a problem, quite the contrary. I am not sure we want to process alerts before system boots and is in a "stable" state. It can also try to send alerts out which requires network.

**Comment by sonicaj** [comment]
In that case, it makes most sense to only have one parameter here as you mentioned above. I'll update the PR, thank you

**Comment by william-gr** [comment]
Lets wait and see what @themylogin thinks too

File: src/middlewared/middlewared/main.py [hunk 1]
```diff
@@ -849,27 +849,51 @@ def sort_key(plugin__function):
 
         self.logger.debug('All plugins loaded')
 
-    def __setup_periodic_tasks(self):
+    async def __setup_periodic_tasks(self):
+        system_ready_tasks = []
+
+        async def execute_periodic_func(delay, method, service_name, service_obj, method_name, interval):
```
Comment (themylogin) [inline src/middlewared/middlewared/main.py]: There are no awaits inside, you can omit async def

File: src/middlewared/middlewared/main.py [hunk 2]
```diff
@@ -849,27 +849,51 @@ def sort_key(plugin__function):
...
+            self.loop.call_later(
+                delay,
+                functools.partial(
+                    self.__call_periodic_task,
+                    method, service_name, service_obj, method_name, interval
+                )
+            )
+
+        async def system_read_task_event_handler(middleware, event_type, args):
+            if args['id'] == 'ready':
```
Comment (themylogin) [inline src/middlewared/middlewared/main.py]: Becomes a pattern, can we do something to allow subscribing to `system.ready`?

**Comment by themylogin** [comment]
I agree with @william-gr, let's just do all run_on_start tasks when system is ready.

File: src/middlewared/middlewared/plugins/smb.py [hunk 3]
```diff
@@ -1144,7 +1144,7 @@ class Config:
                     args=','.join(share['share_acl'].split())
                 )
 
-    @periodic(3600, run_on_start=False)
+    @periodic(3600)
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/smb.py]: Why did you change it here?

**Comment by william-gr** [comment]
@themylogin I agree, should keep it as minimum as possible and merge back to 11.3
